src/setup/rgw.py

```diff
@@ -14,77 +14,219 @@
+            # Load template ceph.conf file and updated required field.
+            kv_list = [
+                ('global>fsid', str(uuid.uuid1())),
+                ('global>mon host', host_ip)]
+            Rgw._update_config(ceph_tmpl_idx, ceph_tmpl_url, kv_list)
+
+            # Copy ceph_conf_tmpl_file to /etc/ceph/ dir.
+            shutil.copyfile(CEPH_CONF_TMPL, CEPH_CONF)
+            Log.info(f'{CEPH_CONF_TMPL} config copied to {CEPH_CONF}.')
+
+            # create admin keyring.
+            cmd = f"sudo ceph-authtool --create-keyring {CEPH_ADMIN_KEYRING} \
+                --gen-key -n client.admin --cap mon 'allow *' --cap osd 'allow *'"
+            _, err, rc, = SimpleProcess(cmd).run()
+            if rc != 0:
+                raise SetupError(errno.EINVAL, f'"{cmd}" failed with error {err}.')
+        except Exception as e:
+            raise SetupError(errno.EINVAL, f'Error ocurred while fetching node ip, {e}')
+
+        Log.info('Prepare phase completed.')
 
         return 0
 
     @staticmethod
-    def config(config_path: str):
-        """
-        Performs configurations.
-        """
+    def config(conf: MappedConf):
+        """Performs configurations."""
+
+        Log.info('Config phase started.')
 
+        # TODO: Inside [client.rgw.<hostname>] section and
+        # add motr,hax endpoints and rgw client related parameters.
+        Log.info('Config phase completed.')
         return 0
 
     @staticmethod
-    def init(config_path: str):
-        """
-        Perform initialization.
-        """
-
+    def init(conf: MappedConf):
+        """Perform initialization."""
+
+        Log.info('Init phase started.')
+        # TODO: Create admin user.
+        # admin user should be created only on one node.
+        # 1. While creating admin user, global lock created in consul kv store.
+        # (rgw_consul_index, cortx>rgw>volatile>rgw_lock, machine_id)
+        # 2. Before creating admin user.
+        #    a. Check for rgw_lock in consul kv store.
+        #    b. Create user only if lock value is None/machine-id.
+
+        rgw_lock = False
+        rgw_lock_key = 'component>rgw>volatile>rgw_lock'
+        rgw_consul_idx = 'rgw_consul_idx'
+        # Get consul url from cortx config.
+        consul_url = Rgw._get_consul_url(conf)
+        # Check for rgw_lock in consul kv store.
+        Log.info('Checking for rgw lock in consul kv store.')
+        Conf.load(rgw_consul_idx, consul_url)
+
+        while(True):
+            try:
+                rgw_lock_val = Conf.get(rgw_consul_idx, rgw_lock_key)
+                Log.info(f'rgw_lock value - {rgw_lock_val}')
+                if rgw_lock_val is None:
+                    Log.info(f'Setting confstore value for key :{rgw_lock_key}'
+                        f' and value as :{Rgw._machine_id}')
+                    Rgw._update_config(
+                        rgw_consul_idx, consul_url, [(rgw_lock_key, Rgw._machine_id)])
```

Comment:
how acquiring this lock can be atomic in nature across nodes? There is possibility of this code running in parallel on multiple nodes, able to find no lock acquired & everyone tries to set it then, and only last update(hostname) would reflect in lock, even though calls from nodes will succeed. 

If this was the logic used by Hare / S3Sserver earlier, its not full proof, there are windows where it can cause issues.
Keep a TODO to explore consul supplied lock https://www.consul.io/commands/lock

Reply:
Added TODO.